src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -138,9 +138,18 @@ public ImmutableArray<TValue> SelectAsArray<TValue, TArg>(Func<TState, TArg, TVa
     }
 
     public TextDocumentStates<TState> AddRange(ImmutableArray<TState> states)
-        => new(_ids.AddRange(states.Select(state => state.Id)),
-               States.AddRange(states.Select(state => KeyValuePairUtil.Create(state.Id, state))),
-               filePathToDocumentIds: null);
+    {
+        using var pooledIds = SharedPools.Default<List<DocumentId>>().GetPooledObject();
+        var ids = pooledIds.Object;
```

Comment:
Worth setting the capacity of this list?

Reply:
I think I'd rather not unless it shows up. The reasons:

1) Looks like it would need to ensure the desired size exceeds the existing capacity before setting the capacity, otherwise it may throw
2) Setting the capacity explicitly avoids the double resizing code in EnsureCapacity. So, without doing that ourselves here, we'd potentially be allocating quite a bit more often.

Reply:
Going to go ahead and merge, let me know if you would like followup on this